lib/spack/spack/solver/asp.py

```diff
@@ -116,8 +116,32 @@ def getter(node):
     version_origin_fields,
 )(**{name: i for i, name in enumerate(version_origin_fields)})
 
-#: Named tuple to contain information on declared versions
-DeclaredVersion = collections.namedtuple("DeclaredVersion", ["version", "idx", "origin"])
+
+class DeclaredVersion(object):
+    def __init__(self, version, idx, origin):
+        if not isinstance(version, spack.version.VersionBase):
+            raise ValueError("Unexpected type for declared version: {0}".format(type(version)))
```

Comment:
Use type annotations for this, since we check with `mypy` now.

```python
def __init__(self, version: VersionBase, idx: int, origin: str):
```

I'm on the fence on whether we also want to keep the dynamic check.  I could go either way -- I think it's useful if external code ever calls this, and also for routines where we don't currently check types.

Reply:
Now added (but dynamic check retained).